### PR TITLE
KOGITO-3716 Allow automatic close of staging repo

### PR DESCRIPTION
**Pull request by radtriste**
https://issues.redhat.com/browse/KOGITO-3716

Dependent PRs:
- https://github.com/kiegroup/kogito-pipelines/pull/91
- https://github.com/kiegroup/kogito-runtimes/pull/865
- https://github.com/kiegroup/optaplanner/pull/1010
- https://github.com/kiegroup/kogito-examples/pull/435
- https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/114

Many thanks for submitting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

**WARNING! Please make sure you are opening your PR against `master` branch!**

- [x] You have read the [contributors guide](https://github.com/kiegroup/kogito-runtimes#contributing-to-kogito)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [ ] Pull Request title contains the target branch if not targeting master: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

### CODE REVIEW

File: Jenkinsfile.deploy [hunk 1]
```diff
@@ -306,6 +347,10 @@ void runMaven(String command, String directory, boolean skipTests = false, List
         mvnCmd += " ${extraArgs}"
     }
     dir(directory) {
-        maven.runMaven(mvnCmd, skipTests, ['-fae'])
+        def options = ["-s ${env.GLOBAL_MAVEN_SETTINGS}", '-fae']
+        if(params.MAVEN_DEPENDENCIES_REPOSITORY){
+            options += ['-Denforcer.skip']
+        }
+        maven.runMaven(mvnCmd, skipTests, options)
```
Comment (radtriste) [inline Jenkinsfile.deploy]:
```suggestion
        def opts = ["-s ${env.GLOBAL_MAVEN_SETTINGS}", '-fae']
        if(params.MAVEN_DEPENDENCIES_REPOSITORY){
            opts += ['-Denforcer.skip']
        }
        maven.runMaven(mvnCmd, skipTests, opts)
```

File: Jenkinsfile.deploy [hunk 2]
```diff
@@ -292,9 +307,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
...
+        }
+        runMaven('clean deploy', directory, true, [], extraArgs)
+    } else {
+        // First deploy locally
+        extraArgs = "-DaltDeploymentRepository=local::default::file://${env.MAVEN_DEPLOY_LOCAL_DIR}/${directory}"
+        runMaven('clean deploy', directory, true, [], extraArgs)
+
+        // Deploy to staging repository and close
+        cmd = 'org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy-staged-repository -DkeepStagingRepositoryOnCloseRuleFailure=true -DstagingProgressTimeoutMinutes=10'
+        cmd += "--projects ${parentProjectName ?: directory}"
```
Comment (radtriste) [inline Jenkinsfile.deploy]:
```suggestion
        cmd += " --projects ${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy [hunk 3]
```diff
@@ -292,9 +307,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
...
+        }
+        runMaven('clean deploy', directory, true, [], extraArgs)
+    } else {
+        // First deploy locally
+        extraArgs = "-DaltDeploymentRepository=local::default::file://${env.MAVEN_DEPLOY_LOCAL_DIR}/${directory}"
+        runMaven('clean deploy', directory, true, [], extraArgs)
+
+        // Deploy to staging repository and close
+        cmd = 'org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy-staged-repository -DkeepStagingRepositoryOnCloseRuleFailure=true -DstagingProgressTimeoutMinutes=10'
+        cmd += " --projects ${parentProjectName ?: directory}"
```
Comment (radtriste) [inline Jenkinsfile.deploy]:
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy [hunk 4]
```diff
@@ -292,9 +307,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
...
+        cmd += " -DserverId=${env.NEXUS_RELEASE_REPOSITORY_ID}" 
+        cmd += " -DrepositoryDirectory=${env.MAVEN_DEPLOY_LOCAL_DIR}/${directory}"
+        cmd += " -DstagingProfileId=${env.NEXUS_STAGING_REPOSITORY_ID}" 
+        cmd += " -DstagingDescription=\"Kogito Runtimes ${getProjectVersion()}\""
+        echo "Run maven cmd $cmd"
+        runMaven(cmd, directory)
+
+        // Promote to build profile
+        cmd = "org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:promote -DbuildPromotionProfileId=${env.NEXUS_BUILD_PROMOTION_PROFILE_ID}"
+        cmd += "--projects ${directory}"
```
Comment (radtriste) [inline Jenkinsfile.deploy]:
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy [hunk 5]
```diff
@@ -88,16 +102,17 @@ pipeline {
...
-                                sh "sed -i 's|<pluginRepositories>|<pluginRepositories><pluginRepository><id>staging</id><name>Staging Repository</name><url>${params.MAVEN_DEPENDENCIES_REPOSITORY}</url><layout>default</layout><snapshots><enabled>true</enabled></snapshots><releases><enabled>true</enabled></releases></pluginRepository>|g' ${MAVEN_SETTINGS_FILE}"
-                                // Done to allow Maven to download release artificats from MAVEN_DEPENDENCIES_REPOSITORY
-                                sh "sed -i 's|external:\\*|external:*,!staging|g' ${MAVEN_SETTINGS_FILE}"
+                            if(params.MAVEN_DEPENDENCIES_REPOSITORY){
+                                depsRepositoryContent = "<id>deps-repo</id><name>Dependencies Repository</name><url>${params.MAVEN_DEPENDENCIES_REPOSITORY}</url><layout>default</layout><snapshots><enabled>true</enabled></snapshots><releases><enabled>true</enabled></releases>"
+                                sh "echo '\n-Denforcer.skip' | tee -a optaplanner/.mvn/maven.config"
```
Comment (radtriste) [inline Jenkinsfile.deploy]:
```suggestion
                                sh "echo '\n-Denforcer.skip' | tee -a .mvn/maven.config"
```

**Comment by radtriste** [comment]
Almost there but testing is blocked by https://issues.redhat.com/browse/KOGITO-3795 ...

**Comment by radtriste** [comment]
Jenkins retest this

**Comment by radtriste** [comment]
Jenkins retest this

File: Jenkinsfile.promote [hunk 6]
```diff
@@ -104,17 +107,17 @@ pipeline {
                 script {
                     dir('kogito-examples-bot') {
                         prepareForPR('kogito-examples')
-                        setupMavenConfig()
-                        maven.mvnVersionsSet(getSnapshotVersion(), true)
-                        updateOptaPlannerVersion(util.getNextVersion(getOptaPlannerVersion(), 'micro'))
+
+                        maven.mvnVersionsSet(getMavenConfig(), getSnapshotVersion(), true)
```
Comment (rsynek) [inline Jenkinsfile.promote]: Here we use `getMavenConfig()` while in `kogito-runtimes` we create a `defaultMavenCmd`. Would be nice to keep consistency across the pipelines.

Btw, I prefer this approach over the `defaultMavenCmd`: the `getMavenConfig()` always returns a new instance. Thus, there is no risk of accidentally changing the `defaultMavenCmd` due to forgetting to call `clone()`.

**Review by rsynek — APPROVED** [review]

File: Jenkinsfile.promote (hunk 6 above)
Comment (radtriste) [reply Jenkinsfile.promote]: Until now, `promote` has the `getMavenConfig` and `deploy` use the `default[...]` var.
Could be changed indeed to `getMavenConfig`. will do that.

**Comment by radtriste** [comment]
@rsynek ready to be merged :)

**Pull request merged**